optuna/integration/_allennlp/pruner.py

```diff
@@ -0,0 +1,298 @@
+            `Registrable.register <https://docs.allennlp.org/master/
+            api/common/registrable/#registrable>`_ in AllenNLP.
+
+            """
+
+            def wrapper(subclass: Any, *args: Any, **kwargs: Any) -> Any:
+                return subclass
+
+            return wrapper
+
+
+_PPID = os.getppid()
+
+"""
+User might want to launch multiple studies that uses `AllenNLPExecutor`.
+Because `AllenNLPExecutor` uses environment variables for communicating
+between a parent process and a child process. A parent process creates a study,
+defines a search space, and a child process trains a AllenNLP model by
+`allennlp.commands.train.train_model`. If multiple processes use `AllenNLPExecutor`,
+the one's configuration could be loaded in the another's configuration.
+To avoid this hazard, we add ID of a parent process to each key of
+environment variables.
+"""
+_PREFIX = "{}_OPTUNA_ALLENNLP".format(_PPID)
+_MONITOR = "{}_MONITOR".format(_PREFIX)
+_PRUNER_CLASS = "{}_PRUNER_CLASS".format(_PREFIX)
+_PRUNER_KEYS = "{}_PRUNER_KEYS".format(_PREFIX)
+_STORAGE_NAME = "{}_STORAGE_NAME".format(_PREFIX)
+_STUDY_NAME = "{}_STUDY_NAME".format(_PREFIX)
+_TRIAL_ID = "{}_TRIAL_ID".format(_PREFIX)
+
+
+def _create_pruner() -> Optional[optuna.pruners.BasePruner]:
+    """Restore a pruner which is defined in `create_study`.
+
+    `AllenNLPPruningCallback` is launched as a sub-process of
+    a main script that defines search spaces.
+    An instance cannot be passed directly from the parent process
+    to its sub-process. For this reason, we set information about
+    pruner as environment variables and load them and
+    re-create the same pruner in `AllenNLPPruningCallback`.
+
+    """
+    pruner_class = os.getenv(_PRUNER_CLASS)
+    if pruner_class is None:
+        return None
+
+    pruner_params = _get_environment_variables_for_pruner()
+    pruner = getattr(optuna.pruners, pruner_class, None)
+
+    if pruner is None:
+        return None
+
+    return pruner(**pruner_params)
+
+
+def _get_environment_variables_for_trial() -> Dict[str, Optional[str]]:
+    return {
+        "study_name": os.getenv(_STUDY_NAME),
+        "trial_id": os.getenv(_TRIAL_ID),
+        "storage": os.getenv(_STORAGE_NAME),
+        "monitor": os.getenv(_MONITOR),
+    }
+
+
+def _get_environment_variables_for_pruner() -> Dict[str, Optional[Union[str, int, float, bool]]]:
+    keys = os.getenv(_PRUNER_KEYS)
+
+    # keys would be empty when `_PRUNER_CLASS` is `NopPruner`
+    if keys is None or keys == "":
+        return {}
+
+    kwargs = {}
+    for key in keys.split(","):
+        key_without_prefix = key.replace("{}_".format(_PREFIX), "")
+        value = os.getenv(key)
+        if value is None:
+            raise ValueError(f"{key} is not found in environment variables.")
+        kwargs[key_without_prefix] = eval(value)
+
+    return kwargs
+
+
+def dump_best_config(input_config_file: str, output_config_file: str, study: optuna.Study) -> None:
```

Comment:
The content of this function seems to be completely the same as `_allennlp/dump_best_config.py`. I think we should remove this one or the other.

Reply:
8eede83 removes it.